src/avro.erl

```diff
@@ -309,6 +310,143 @@ decode(avro_json, JSON, TypeOrName, StoreOrLkup, Hook) ->
+      {not_compatible, lists:flatten(RPath), lists:flatten(WPath)}
+  end.
+
+desc(RecordType) when ?IS_RECORD_TYPE(RecordType) ->
+  [{record, avro:get_type_name(RecordType)}];
+desc(ArrayType) when ?IS_ARRAY_TYPE(ArrayType) ->
+  [ItemsType] =  desc(avro_array:get_items_type(ArrayType)),
+  [{avro:get_type_name(ArrayType), ItemsType}];
+desc(MapType) when ?IS_MAP_TYPE(MapType) ->
+  [ItemsType] =  desc(avro_map:get_items_type(MapType)),
+  [{avro:get_type_name(MapType), ItemsType}];
+desc(EnumType) when ?IS_ENUM_TYPE(EnumType) ->
+  [{enum, avro:get_type_name(EnumType)}];
+desc(FixedType) when ?IS_FIXED_TYPE(FixedType) ->
+  [{fixed, avro:get_type_name(FixedType)}];
+desc(UnionType) when ?IS_UNION_TYPE(UnionType) ->
+  [{union, avro:get_type_name(UnionType)}];
+desc(AvroType) ->
+  [avro:get_type_name(AvroType)].
+
+do_is_compatible(Reader, Writer)
+  when ?IS_RECORD_TYPE(Reader) andalso ?IS_RECORD_TYPE(Writer) ->
+  SameType = avro:is_same_type(Reader, Writer),
+  ReaderTypes = avro_record:get_all_field_data(Reader),
+  WriterTypes = avro_record:get_all_field_data(Writer),
+  FieldsDecodable =
+    try
+      lists:all(
+        fun({FieldName, FieldType, Default}) ->
+            case lists:keysearch(FieldName, 1, WriterTypes) of
+              false ->
+                Default =/= ?NO_VALUE;
+            {value, {_, WriterType, _}} ->
+                do_is_compatible(FieldType, WriterType)
+            end
+        end,
+        ReaderTypes
+       )
+    catch throw:{not_compatible, RPath, WPath} ->
+        erlang:throw({ not_compatible, [desc(Reader) | RPath]
+                     , [desc(Writer) | WPath]})
+    end,
+    Result = SameType andalso FieldsDecodable,
+    case Result of
+      true ->
+        true;
+      false ->
+        erlang:throw({not_compatible, desc(Reader), desc(Writer)})
+    end;
+do_is_compatible(Reader, Writer)
+  when ?IS_ENUM_TYPE(Reader) andalso ?IS_ENUM_TYPE(Writer) ->
+  Result = avro:is_same_type(Reader, Writer) andalso
+    Writer#avro_enum_type.symbols -- Reader#avro_enum_type.symbols == [],
+  case Result of
+    true ->
+      true;
+    false ->
+      erlang:throw({not_compatible, desc(Reader), desc(Writer)})
+  end;
+do_is_compatible(Reader, Writer)
+  when ?IS_FIXED_TYPE(Reader) andalso ?IS_FIXED_TYPE(Writer) ->
+  Result = avro:is_same_type(Reader, Writer)
+    andalso (Reader#avro_fixed_type.size == Writer#avro_fixed_type.size),
+  case Result of
+    true ->
+      true;
+    false ->
+      erlang:throw({not_compatible, desc(Reader), desc(Writer)})
+  end;
+do_is_compatible(Reader, Writer)
+  when ?IS_UNION_TYPE(Reader) andalso ?IS_UNION_TYPE(Writer) ->
+  ReaderTypes = avro_union:get_types(Reader),
+  WriterTypes = avro_union:get_types(Writer),
+  CheckFun =
+    fun F([], _) ->
+        true;
+        F([WriterType | WT], [ReaderType | RT]) ->
+        do_is_compatible(ReaderType, WriterType)
+          andalso F(WT, RT)
+    end,
+    try
+      CheckFun(WriterTypes, ReaderTypes)
```

Comment:
nit: on all the places Reader schema goes as 1st argument and writer as 2nd, except this one

Reply:
Reason being that the reader union might have more members than the writer's union.